### PR TITLE
Get Spark working

### PR DESCRIPTION
**Pull request by kayousterhout**
This series of commits gets Spark working.  After these commits, a user can log into the Spark master, start a shell, and successfully run a shuffle job.

### CODE REVIEW

File: spark.js [hunk 1]
```diff
@@ -17,8 +19,16 @@ function setImage(newImage) {
  * @param {number} nWorker The number of workers to boot.
  */
 function Spark(nWorker) {
-  this.master = new Container('spark-ms', image,
-      { command: ['run', 'master'] });
+  // Set a spark-env.sh file on each machine, which will be sourced before starting
+  // any Spark processes.
+  const sparkEnvContent = fs.readFileSync(path.join(__dirname, 'spark-env.sh'),
+    { encoding: 'utf8' });
+  const sparkEnvFilepathToContent = { '/spark/conf/spark-env.sh': sparkEnvContent };
```
Comment (kayousterhout) [inline spark.js]: BTW you may be wondering why this is added here instead of in the Dockerfile.  This is the way I expect to encourage users to change their Spark configuration going forward (by making changes to this file), so it's better to include it in the blueprint rather than as part of the image.

**Comment by ejj** [comment]
I'd like to review this too.

File: spark.js [hunk 2]
```diff
@@ -40,9 +40,16 @@ function Spark(nWorker) {
...
   this.exposeUIToPublic = function exposeUIToPublic() {
+    // Expose the Standalone master UI (which shows all of the workers and all of
+    // the applications that have run).
     allow(publicInternet, this.master, 8080);
     allow(publicInternet, this.workers, 8081);
 
+    // Expose the per-job UI (which shows each application).
+    // This will only work when there's only one job; otherwise, the UI will use
+    // increasing port numbers for the other jobs, and those will not be accessible.
+    allow(publicInternet, this.master, 4040);
```
Comment (ejj) [inline spark.js]: So if I start a job, and then it finishes, and then a new one starts will that new one re-use 4040?

We support port ranges for allow don't we?  I wonder if we should just allow 4040-4100 or something.   It's a hack but I suspect it's good enough.

File: Dockerfile [hunk 3]
```diff
@@ -4,11 +4,13 @@ Maintainer Ethan J. Jackson
...
         default-jre-headless \
         python-minimal \
+	# wget is used to download Spark, and is used once the container starts to
+	# get the container's public IP address by reading checkip.amazonaws.com.
+	# This can be removed once Kelda has a mechanism to set an environment variable
+	# in the container to the container's public IP address.
         wget \
 && wget -qO- http://www-us.apache.org/dist/spark/spark-2.2.0/spark-2.2.0-bin-hadoop2.7.tgz | tar -xzf - \
 && mv /spark* /spark \
-&& apt-get remove --purge -y wget \
```
Comment (ejj) [inline Dockerfile]: I found this sentence a bit hard to parse: "Right now, for the user to
run a job and correctly set this, the user would need to manually
set the SPARK_PUBLIC_DNS environment variable before running a spark-submit
command (or run-example or spark-shell) to submit a job."

File: spark-env.sh [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# Set Spark's public IP address so that the links in the UI work.
+export SPARK_PUBLIC_DNS=`wget -q -O - 2>&1 http://checkip.amazonaws.com`
```
Comment (ejj) [inline spark-env.sh]: Would you add an XXX comment to this saying that this file should be generated by spark.js once we have the required functionality?

File: spark.js (hunk 1 above)
Comment (ejj) [reply spark.js]: Fair enough =)

File: spark.js (hunk 2 above)
Comment (kayousterhout) [reply spark.js]: A new one will use 4040; this is only an issue if multiple jobs are running concurrently (which requires some additional setup -- like configuring the resources each job should get -- that we don't directly support currently).  Do you still think we should use a range?

Comment (ejj) [reply spark.js]: I don't really see a reason why not to I guess.  But maybe the range can be smaller.  Do people ever run more than 10 jobs concurrently?  Maybe just 4040-4050.

I don't really see a security risk for opening the ports, and I see a very small but non zero probability of someone running into the issue.

All of this assumes that we still have the ability to use port ranges in allow.  If that's not longer then case, then I agree, it's probably not worth it.

File: spark-env.sh (hunk 4 above)
Comment (ejj) [reply spark-env.sh]: Or maybe the file shouldn't be generated, but the value of SPARK_PUBLIC_DNS should be ...

File: spark.js (hunk 2 above)
Comment (kayousterhout) [reply spark.js]: I realized that I don't think multiple concurrent jobs will work because of the other two ports that we configure -- the two jobs will try to, for example, run a block manager on the same port, which will cause one to fail.  I added to our internal checklist an item to add support for multiple concurrent jobs (and I'll do this in a later PR).

**Comment by kayousterhout** [comment]
@ejj updated!
